src/vs/platform/terminal/node/terminalEnvironment.ts

```diff
@@ -72,3 +75,175 @@ export async function findExecutable(command: string, cwd?: string, paths?: stri
 	const fullPath = path.join(cwd, command);
 	return await exists(fullPath) ? fullPath : undefined;
 }
+
+export interface IShellIntegrationInjection {
+	/**
+	 * A new set of arguments to use.
+	 */
+	newArgs: string[] | undefined;
+	/**
+	 * An optional environment to mixing to the real environment.
+	 */
+	envMixin?: IProcessEnvironment;
+	/**
+	 * An optional array of files to copy from `source` to `dest`.
+	 */
+	copyFiles?: {
+		source: string;
+		dest: string;
+	}[];
+}
+
+/**
+ * For a given shell launch config, returns arguments to replace and an optional environment to
+ * mixin to the SLC's environment to enable shell integration. This must be run within the context
+ * that creates the process to ensure accuracy. Returns undefined if shell integration cannot be
+ * enabled.
+ */
+export function getShellIntegrationInjection(
+	shellLaunchConfig: IShellLaunchConfig,
+	options: ITerminalProcessOptions['shellIntegration']
+): IShellIntegrationInjection | undefined {
+	// Shell integration arg injection is disabled when:
+	// - The global setting is disabled
+	// - There is no executable (not sure what script to run)
+	// - The terminal is used by a feature like tasks or debugging
+	if (!options.enabled || !shellLaunchConfig.executable || shellLaunchConfig.isFeatureTerminal) {
+		return undefined;
+	}
+
+	const originalArgs = shellLaunchConfig.args;
+	const shell = path.basename(shellLaunchConfig.executable).toLowerCase();
+	const appRoot = path.dirname(FileAccess.asFileUri('', require).fsPath);
+	let newArgs: string[] | undefined;
+
+	// Windows
+	if (isWindows) {
+		if (shell === 'pwsh.exe') {
+			if (!originalArgs || arePwshImpliedArgs(originalArgs)) {
+				newArgs = shellIntegrationArgs.get(ShellIntegrationExecutable.WindowsPwsh);
+			} else if (arePwshLoginArgs(originalArgs)) {
+				newArgs = shellIntegrationArgs.get(ShellIntegrationExecutable.WindowsPwshLogin);
+			}
+			if (!newArgs) {
+				return undefined;
+			}
+			if (newArgs) {
+				const additionalArgs = options.showWelcome ? '' : ' -HideWelcome';
+				newArgs = [...newArgs]; // Shallow clone the array to avoid setting the default array
+				newArgs[newArgs.length - 1] = format(newArgs[newArgs.length - 1], appRoot, additionalArgs);
+			}
+			return { newArgs };
+		}
+		return undefined;
+	}
+
+	// Linux & macOS
+	const envMixin: IProcessEnvironment = {};
+	const copyFiles: IShellIntegrationInjection['copyFiles'] = [];
```

Comment:
can this move into the block for `zsh` so it's clear that's when this gets used?